enterprise/coderd/prebuilds/reconcile.go

```diff
@@ -0,0 +1,511 @@
+
+	ticker := c.clock.NewTicker(reconciliationInterval)
+	defer ticker.Stop()
+	defer func() {
+		c.done <- struct{}{}
+	}()
+
+	// nolint:gocritic // Reconciliation Loop needs Prebuilds Orchestrator permissions.
+	ctx, cancel := context.WithCancelCause(dbauthz.AsPrebuildsOrchestrator(ctx))
+	c.cancelFn = cancel
+
+	for {
+		select {
+		// TODO: implement pubsub listener to allow reconciling a specific template imperatively once it has been changed,
+		//		 instead of waiting for the next reconciliation interval
+		case <-ticker.C:
+			// Trigger a new iteration on each tick.
+			err := c.ReconcileAll(ctx)
+			if err != nil {
+				c.logger.Error(context.Background(), "reconciliation failed", slog.Error(err))
+			}
+		case <-ctx.Done():
+			// nolint:gocritic // it's okay to use slog.F() for an error in this case
+			// because we want to differentiate two different types of errors: ctx.Err() and context.Cause()
+			c.logger.Warn(context.Background(), "reconciliation loop exited", slog.Error(ctx.Err()), slog.F("cause", context.Cause(ctx)))
+			return
+		}
+	}
+}
+
+func (c *StoreReconciler) Stop(ctx context.Context, cause error) {
+	if cause != nil {
+		c.logger.Error(context.Background(), "stopping reconciler due to an error", slog.Error(cause))
+	} else {
+		c.logger.Info(context.Background(), "gracefully stopping reconciler")
+	}
+
+	if c.isStopped() {
+		return
+	}
+	c.stopped.Store(true)
+	if c.cancelFn != nil {
+		c.cancelFn(cause)
+	}
+
+	select {
+	// Give up waiting for control loop to exit.
+	case <-ctx.Done():
+		// nolint:gocritic // it's okay to use slog.F() for an error in this case
+		// because we want to differentiate two different types of errors: ctx.Err() and context.Cause()
+		c.logger.Error(context.Background(), "reconciler stop exited prematurely", slog.Error(ctx.Err()), slog.F("cause", context.Cause(ctx)))
+	// Wait for the control loop to exit.
+	case <-c.done:
+		c.logger.Info(context.Background(), "reconciler stopped")
+	}
+}
+
+func (c *StoreReconciler) isStopped() bool {
+	return c.stopped.Load()
+}
+
+// ReconcileAll will attempt to resolve the desired vs actual state of all templates which have presets with prebuilds configured.
+//
+// NOTE:
+//
+// This function will kick of n provisioner jobs, based on the calculated state modifications.
+//
+// These provisioning jobs are fire-and-forget. We DO NOT wait for the prebuilt workspaces to complete their
+// provisioning. As a consequence, it's possible that another reconciliation run will occur, which will mean that
+// multiple preset versions could be reconciling at once. This may mean some temporary over-provisioning, but the
+// reconciliation loop will bring these resources back into their desired numbers in an EVENTUALLY-consistent way.
+//
+// For example: we could decide to provision 1 new instance in this reconciliation.
+// While that workspace is being provisioned, another template version is created which means this same preset will
```

Comment:
I think we should clarify what "this same preset" means a little more precisely, since the preset ID will strictly be different for new template versions.